### PR TITLE
Add support for filtering on node labels

### PR DESCRIPTION
**Pull request by anshulpundir**
Full diff: https://github.com/docker/swarmkit/compare/8852e8840e30d69db0b39a4a3d6447362e17c64f...496d19be63751d60a55887604683e5cb1a5541aa

Relevant changes:

- https://github.com/docker/swarmkit/pull/2662 add filtering on node labels
  - A rewrite of https://github.com/docker/swarmkit/pull/2062.
  - Relates to https://github.com/moby/moby/issues/27231, https://github.com/moby/moby/issues/28209

ping @andrewhsu @thaJeztah @dperny

### CODE REVIEW

**Review by yongtang — APPROVED** [review]
LGTM once janky is 💚

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Looks like we need some changes to make the new filter work;

- update `newListNodesFilters()` to accept the new filter https://github.com/moby/moby/blob/3a633a712c8bbb863fe7e57ec132dd87a9c4eff7/daemon/cluster/filters.go#L12-L46
- update the Swagger file to document the newly accepted filter; https://github.com/moby/moby/blob/bd06a5ea4df919ff323510fba10801b5673a3af6/api/swagger.yaml?utf8=✓#L8577-L8589
- add a note to the API changelog to mention the new filter; https://github.com/moby/moby/blob/master/docs/api/version-history.md. This change will be in API 1.39, so either a new entry should be added for that version, or that would come with https://github.com/moby/moby/pull/37472 (once merged)

**Comment by anshulpundir** [comment]
addressed comments from @thaJeztah

File: daemon/cluster/filters.go
Comment (thaJeztah) [inline daemon/cluster/filters.go]: looks like this should be singular

File: docs/api/version-history.md
Comment (thaJeztah) [inline docs/api/version-history.md]: typo s/ilter/filter/

Comment (thaJeztah) [reply docs/api/version-history.md]: also s/label/node-label/

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml]: also thinking if this should be `node.label` (full stop, instead of hyphen), as was proposed in https://github.com/moby/moby/issues/28209

ping @vdemeester @dperny?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by andrewhsu** [comment]
I'd suggest ignoring the z [errors](https://jenkins.dockerproject.org/job/Docker-PRs-s390x/10775/consoleFull) because we've had intermittent networking issues with our z infrastructure.
```
Downloading 'library/busybox:glibc@sha256:0b55a30394294ab23b9afd58fab94e61a923f5834fba7ddbae7f8e0c11ba85e6' (1 layers)...
curl: (28) Operation timed out after 300521 milliseconds with 0 out of 0 bytes received
```

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml]: hm, looks like you changed the example value, instead of the "key"; we could discuss quickly, but I think the idea in https://github.com/moby/moby/issues/28209 was to have `--filter node.label=foobar` (so `node.label` instead of `node-label`)

@vdemeester ^^

File: daemon/cluster/filters.go
Comment (thaJeztah) [inline daemon/cluster/filters.go]: same here (and other places below)

File: docs/api/version-history.md
Comment (thaJeztah) [inline docs/api/version-history.md]: The examples here still don't match; e.g. `label=<key>=<value>` should be either `node-label=<key>=<value>` or `node.label=<key>=<value>`

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: api/swagger.yaml
Comment (vdemeester) [reply api/swagger.yaml]: Yeah, to have `node.label`, `engine.label` and `label`. The dot `.` was maintly to be consistent with constraint if I rembember correctly :angel:

Comment (thaJeztah) [reply api/swagger.yaml]: Ah; yes; that was the reason

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Review by andrewhsu — APPROVED** [review]
LGTM

ignore the z [error](https://jenkins.dockerproject.org/job/Docker-PRs-s390x/10790/console):
```
curl: (56) SSL read: error:00000000:lib(0):func(0):reason(0), errno 104
```
because z infrastructure issues

**Comment by thaJeztah** [comment]
Failure on experimental (think that's a flaky)

```
19:17:48 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection
19:17:48 
19:17:48 [dfbd0fadafb65] waiting for daemon to start
19:17:48 [dfbd0fadafb65] daemon started
19:17:48 
19:17:48 [d96abff1621cb] waiting for daemon to start
19:17:48 [d96abff1621cb] daemon started
19:17:48 
19:17:48 [df10c3972bb8f] waiting for daemon to start
19:17:48 [df10c3972bb8f] daemon started
19:17:48 
19:17:48 [dfbd0fadafb65] exiting daemon
19:17:48 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
19:17:48 [d96abff1621cb] exiting daemon
19:17:48 [df10c3972bb8f] exiting daemon

**Comment by thaJeztah** [comment]
Dropped the swarmkit bump commit, because master is now ahead of that commit, and fixed the merge-conflict in  docs/api/version-history.md

**Comment by thaJeztah** [comment]
Argh. Janky failing again on that swagger-check; https://github.com/moby/moby/pull/36714. I'll restart, but we can likely look at the "experimental" build as well (which should run the same tests)

**Comment by anshulpundir** [comment]
thx taking care of this! @thaJeztah

**Pull request merged**
